File: 366-find-leaves-of-binary-tree/366-find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        def getNodeHeights(node: TreeNode, h_dict, h_list):
            if node is None:
                return 0
            height = 1 + max(getNodeHeights(node.left, h_dict, h_list), getNodeHeights(node.right, h_dict, h_list))
            if height in h_dict:
                h_dict[height].append(node.val)
            else:
                h_dict[height] = [node.val]
                h_list.append(height)
            return height
        h_dict = dict()
        h_list = []
        getNodeHeights(root, h_dict, h_list)
        return [h_dict[h] for h in h_list]
        

File: 366-find-leaves-of-binary-tree/366-find-leaves-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        findHeight(root, res);
        return res;
    }
    private int findHeight(TreeNode node, List<List<Integer>> res){
        if(node==null)return -1;
        int leftHeight = findHeight(node.left, res);
        int rightHeight = findHeight(node.right, res);
        int curHeight = 1+Math.max(leftHeight, rightHeight);
        if(res.size()==curHeight){
            res.add(new ArrayList<>());
        }
        res.get(curHeight).add(node.val);
        return curHeight;
    }
}